### PR TITLE
Make recovery more reliable

### PR DESCRIPTION
**Pull request by jsteemann**
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2079/

### CODE REVIEW

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1997/

File: arangod/RocksDBEngine/RocksDBCollectionMeta.cpp [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+ * specified transaction identifier, then this will simply do nothing.
+ *
+ * @param trxId Identifier for active transaction (should match input to
+ *              earlier `placeBlocker` call)
+ */
+void RocksDBCollectionMeta::removeBlocker(uint64_t trxId) {
+  WRITE_LOCKER(locker, _blockerLock);
+  auto it = _blockers.find(trxId);
+  if (ADB_LIKELY(_blockers.end() != it)) {
+    auto cross = _blockersBySeq.find(std::make_pair(it->second, it->first));
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollectionMeta.cpp]: you could just use `_blockersBySeq.erase(...)` instead of looking up the value and then deleting it.
the result will be the same.

File: arangod/RocksDBEngine/RocksDBCollectionMeta.cpp [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+    TRI_ASSERT(_blockersBySeq.end() != cross);
+    if (ADB_LIKELY(_blockersBySeq.end() != cross)) {
+      _blockersBySeq.erase(cross);
+    }
+    _blockers.erase(it);
+  }
+}
+
+/// @brief updates and returns the largest safe seq to squash updated against
+rocksdb::SequenceNumber RocksDBCollectionMeta::committableSeq() {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollectionMeta.cpp]: can be made const
```suggestion
rocksdb::SequenceNumber RocksDBCollectionMeta::committableSeq() const {
```

File: arangod/RocksDBEngine/RocksDBCollectionMeta.h [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+   * updates in case of successful commit. If no blocker exists with the
+   * specified transaction identifier, then this will simply do nothing.
+   *
+   * @param trxId Identifier for active transaction (should match input to
+   *              earlier `placeBlocker` call)
+   */
+  void removeBlocker(uint64_t trxId);
+  
+  /// @brief updates and returns the largest safe seq to squash updated against
+  rocksdb::SequenceNumber committableSeq();
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollectionMeta.h]: can be made const
```suggestion
  rocksdb::SequenceNumber committableSeq() const;
```

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/3050/

File: arangod/RocksDBEngine/RocksDBCollectionMeta.cpp [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+      << "type '" << idx->typeName() << "' with id '" << idx->id() << "'";
+      idx->recalculateEstimates();
+      continue;
+    }
+    
+    StringRef estimateInput(value.data() + sizeof(uint64_t),
+                            value.size() - sizeof(uint64_t));
+    
+    uint64_t committedSeq = rocksutils::uint64FromPersistent(value.data());
+    if (RocksDBCuckooIndexEstimator<uint64_t>::isFormatSupported(estimateInput)) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollectionMeta.cpp]: what do we do if the format is not supported? is this supposed to happen at all?

File: arangod/RocksDBEngine/RocksDBCollectionMeta.cpp [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+      auto est = std::make_unique<RocksDBCuckooIndexEstimator<uint64_t>>(committedSeq, estimateInput);
+      LOG_TOPIC(ERR, Logger::ENGINES)
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollectionMeta.cpp]: does this situation justify an ERR log level?

File: arangod/RocksDBEngine/RocksDBCollectionMeta.cpp [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+  rocksdb::WriteOptions wo;
+  wo.sync = true;
+  
+  
+  // Step 1. delete the document count
+  RocksDBKey key;
+  key.constructCounterValue(objectId);
+  rocksdb::Status s =db->Delete(wo, cf, key.string());
+  if (!s.ok()) {
+    return rocksutils::convertStatus(s);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollectionMeta.cpp]: shouldn't we just go on deleting other keys for the collection? otherwise even more "garbage" will remain.

File: arangod/RocksDBEngine/RocksDBCollectionMeta.cpp [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+  
+  rocksdb::ColumnFamilyHandle* const cf = RocksDBColumnFamily::definitions();
+  rocksdb::WriteOptions wo;
+  wo.sync = true;
+  
+  RocksDBKey key;
+  key.constructIndexEstimateValue(objectId);
+  rocksdb::Status s = db->Delete(wo, cf, key.string());
+  if (!s.ok() && !s.IsNotFound()) {
+    return rocksutils::convertStatus(s);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollectionMeta.cpp]: shouldn't we just go on deleting the estimates for other remaining indexes of the collection? otherwise even more "garbage" will remain.

File: arangod/RocksDBEngine/RocksDBCollectionMeta.h [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  /// @brief apply counter adjustments, only call from sync thread
+  rocksdb::SequenceNumber applyAdjustments(rocksdb::SequenceNumber commitSeq,
+                                           bool& didWork);
+  
+private:
+
+  // TODO we should probably use flat_map or abseils Swiss Tables
+  
+  mutable arangodb::basics::ReadWriteSpinLock _blockerLock;
+  /// @brief blocker identifies a transaction beeing committet
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollectionMeta.h]:
```suggestion
  /// @brief blocker identifies a transaction being committed
```

File: arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h [hunk 9]
```diff
@@ -287,34 +292,30 @@ class RocksDBCuckooIndexEstimator {
...
-      for (uint64_t i = 0; i < (_size * _counterSize * SlotsPerBucket);
-           i += _counterSize) {
+      for (uint64_t i = 0; i < (_size * kCounterSize * kSlotsPerBucket);
+           i += kCounterSize) {
         rocksutils::uint32ToPersistent(
             serialized, *(reinterpret_cast<uint32_t*>(_counters + i)));
       }
 
-      bool havePendingUpdates = !_blockers.empty() || !_insertBuffers.empty() ||
+      bool havePendingUpdates = /*!_blockers.empty() ||*/ !_insertBuffers.empty() ||
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h]: why is this part commented out?

File: arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h [hunk 10]
```diff
@@ -475,58 +476,58 @@ class RocksDBCuckooIndexEstimator {
...
-    if (ADB_LIKELY(_blockers.end() != it)) {
-      auto cross = _blockersBySeq.find(std::make_pair(it->second, it->first));
-      TRI_ASSERT(_blockersBySeq.end() != cross);
-      if (ADB_LIKELY(_blockersBySeq.end() != cross)) {
-        _blockersBySeq.erase(cross);
-      }
-      _blockers.erase(it);
-    }
-  }
+//
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h]: can we remove this entire part if it is unneeded?

File: arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h [hunk 11]
```diff
@@ -649,22 +660,20 @@ class RocksDBCuckooIndexEstimator {
...
-    // if we have a blocker with a lower value than current, compare it
-    if (!_blockersBySeq.empty()) {
-      auto it = _blockersBySeq.begin();
-      minSeq = std::min(minSeq, it->first);
-    }
-
-    return minSeq;
-  }
+//  /// @brief updates and returns the largest safe seq to consider committed
+//  rocksdb::SequenceNumber committableSeq() {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h]: can we remove this part?

File: arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h [hunk 12]
```diff
@@ -972,35 +981,32 @@ class RocksDBCuckooIndexEstimator {
...
   uint64_t _nrTotal;    // number of elements included in total (not cuckood)
-  unsigned _maxRounds;  // maximum number of cuckoo rounds on insertion
 
-  rocksdb::SequenceNumber mutable _committedSeq;
+  std::atomic<rocksdb::SequenceNumber> _committedSeq;
   std::atomic<bool> _needToPersist;
 
-  std::map<uint64_t, rocksdb::SequenceNumber> _blockers;
-  std::set<std::pair<rocksdb::SequenceNumber, uint64_t>> _blockersBySeq;
+//  std::map<uint64_t, rocksdb::SequenceNumber> _blockers;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCuckooIndexEstimator.h]: remove?

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 13]
```diff
@@ -1259,25 +1256,49 @@ arangodb::Result RocksDBEngine::dropCollection(
...
   // Now Collection is gone.
   // Cleanup data-mess
 
-  // Unregister counter
-  _settingsManager->removeCounter(coll->objectId());
-
+  // Unregister collection metadata
+  Result res = RocksDBCollectionMeta::deleteCollectionMeta(_db, coll->objectId());
+  if (res.fail()) {
+    return res;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBEngine.cpp]: if this fails, shouldn't we go on with the deletion?

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 14]
```diff
@@ -1259,25 +1256,49 @@ arangodb::Result RocksDBEngine::dropCollection(
...
   }
+  
+  // delete indexes, RocksDBIndex::drop() has its own check
+  std::vector<std::shared_ptr<Index>> vecShardIndex = coll->getIndexes();
+  TRI_ASSERT(!vecShardIndex.empty());
+  for (auto& index : vecShardIndex) {
+    RocksDBIndex* ridx = static_cast<RocksDBIndex*>(index.get());
+    res = RocksDBCollectionMeta::deleteIndexEstimate(_db, ridx->objectId());
+    if (res.fail()) {
+      return res;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBEngine.cpp]: if this fails, shouldn't we go on deleting?

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 15]
```diff
@@ -1259,25 +1256,49 @@ arangodb::Result RocksDBEngine::dropCollection(
...
+    }
+    
+    int dropRes = index->drop();
+    if (dropRes != TRI_ERROR_NO_ERROR) {
+      // We try to remove all indexed values.
+      // If it does not work they cannot be accessed any more and leaked.
+      // User View remains consistent.
+      LOG_TOPIC(ERR, Logger::ENGINES) << "unable to drop index: "
+      << TRI_errno_string(dropRes);
+      return TRI_ERROR_NO_ERROR;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBEngine.cpp]: if it fails here, shall we go on deleting?

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 16]
```diff
@@ -1781,6 +1789,11 @@ Result RocksDBEngine::dropDatabase(TRI_voc_tick_t id) {
         // delete index documents
         uint64_t objectId =
             basics::VelocyPackHelper::stringUInt64(it, "objectId");
+        res = RocksDBCollectionMeta::deleteIndexEstimate(_db, objectId);
+        if (res.fail()) {
+          return;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBEngine.cpp]: shouldn't we go on deleting here?

File: arangod/RocksDBEngine/RocksDBIndex.cpp [hunk 17]
```diff
@@ -359,6 +341,7 @@ RocksDBCuckooIndexEstimator<uint64_t>* RocksDBIndex::estimator() {
   return nullptr;
 }
 
-bool RocksDBIndex::needToPersistEstimate() const {
-  return false;
+void RocksDBIndex::setEstimator(std::unique_ptr<RocksDBCuckooIndexEstimator<uint64_t>>) {
+  // Nothing to do.
+  return;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBIndex.cpp]: no return statement needed here
```suggestion
```

File: arangod/RocksDBEngine/RocksDBSettingsManager.cpp [hunk 18]
```diff
@@ -53,47 +53,98 @@
...
-    << "writing counter for collection with objectId '" << pair.first << "' failed: "
-      << s.ToString();
-    return std::make_pair(convertStatus(s), returnSeq);
-  }
-
-  returnSeq = std::min(returnSeq, pair.second._sequenceNum);
-  return std::make_pair(Result(), returnSeq);
-}
-}  // namespace
+//namespace {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBSettingsManager.cpp]: remove all this?

File: arangod/RocksDBEngine/RocksDBSettingsManager.cpp [hunk 19]
```diff
@@ -423,73 +237,69 @@ Result RocksDBSettingsManager::sync(bool force) {
...
+  
+  if (rtrx->GetNumKeys() == 0) {
+    LOG_DEVEL << "nothing happened skipping settings up to " << minSeqNr;
+    return Result(); // nothing was written
   }
-
-  Result res = writeSettings(rtrx.get(), _builder, seqNumber);
-
+  
+//  for (std::pair<uint64_t, CMValue> const& pair : countersCpy) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBSettingsManager.cpp]: remove?

File: arangod/RocksDBEngine/RocksDBSettingsManager.cpp [hunk 20]
```diff
@@ -538,7 +348,7 @@ void RocksDBSettingsManager::loadSettings() {
         if (slice.hasKey("lastSync")) {
           _lastSync =
               basics::VelocyPackHelper::stringUInt64(slice.get("lastSync"));
-          LOG_TOPIC(TRACE, Logger::ENGINES)
+          LOG_TOPIC(ERR, Logger::ENGINES)
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBSettingsManager.cpp]: IMHO this should not have log level ERR
```suggestion
          LOG_TOPIC(TRACE, Logger::ENGINES)
```

File: arangod/RocksDBEngine/RocksDBTransactionCollection.cpp [hunk 21]
```diff
@@ -280,33 +280,41 @@ void RocksDBTransactionCollection::addOperation(
...
-    if (idx == nullptr) {
-      TRI_ASSERT(false); // Index reported estimates, but does not exist
-      continue;
-    }
-    auto ridx = static_cast<RocksDBIndex*>(idx.get());
-    auto estimator = ridx->estimator();
-    if (estimator) {
-      estimator->placeBlocker(trxId, preCommitSeq);
-    }
+//  for (auto const& pair : _trackedIndexOperations) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBTransactionCollection.cpp]: remove?

File: arangod/RocksDBEngine/RocksDBTransactionCollection.cpp [hunk 22]
```diff
@@ -280,33 +280,41 @@ void RocksDBTransactionCollection::addOperation(
...
-    if (idx == nullptr) {
-      TRI_ASSERT(false); // Index reported estimates, but does not exist
-      continue;
-    }
-    auto ridx = static_cast<RocksDBIndex*>(idx.get());
-    auto estimator = ridx->estimator();
-    if (estimator) {
-      estimator->removeBlocker(trxId);
-    }
+//  for (auto const& pair : _trackedIndexOperations) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBTransactionCollection.cpp]: remove?

File: arangod/RocksDBEngine/RocksDBTransactionCollection.cpp [hunk 23]
```diff
@@ -317,33 +325,35 @@ void RocksDBTransactionCollection::commitCounts(uint64_t trxId,
   // Update the collection count
   int64_t const adjustment = _numInserts - _numRemoves;
   if (commitSeq != 0) { // is '0' for filling new indexes
-    if (_numInserts != 0 || _numRemoves != 0 || _revision != 0) {
+    if (hasOperations()) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBTransactionCollection.cpp]: the previous code was only checking for inserts and removes, but `hasOperations()` will also check for updates:
```
122 bool RocksDBTransactionCollection::hasOperations() const {
123   return (_numInserts > 0 || _numRemoves > 0 || _numUpdates > 0);
124 }
```
Intentional?

Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBTransactionCollection.cpp]: Furthermore, we are not explicitly checking `_revision` anymore here, just the number of modifications.

File: arangod/RocksDBEngine/RocksDBVPackIndex.cpp [hunk 24]
```diff
@@ -1282,9 +1234,33 @@ RocksDBCuckooIndexEstimator<uint64_t>* RocksDBVPackIndex::estimator() {
...
+  TRI_ASSERT(_estimator != nullptr);
+  _estimator->clear();
+  
+  rocksdb::TransactionDB* db = rocksutils::globalRocksDB();
+  rocksdb::SequenceNumber seq = db->GetLatestSequenceNumber();
+  
+  auto bounds = getBounds();
+  rocksdb::Slice const end = bounds.end();
+  rocksdb::ReadOptions options;
+  options.iterate_upper_bound = &end;  // save to use on rocksb::DB directly
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBVPackIndex.cpp]:
```suggestion
  options.iterate_upper_bound = &end;  // safe to use on rocksb::DB directly
```

File: arangod/StorageEngine/PhysicalCollection.h [hunk 25]
```diff
@@ -107,8 +107,17 @@ class PhysicalCollection {
...
   std::vector<std::shared_ptr<Index>> getIndexes() const;
+  template<typename F>
+  void enumerateIndexes(F&& f) {
+    _indexesLock.readLock(); // avoid including ReadLocker.h
+    try {
+      for (auto& idx : _indexes) {
+        std::forward<F>(f)(idx);
+      }
+    } catch(...) {}
+    _indexesLock.unlockRead();
```
Comment (jsteemann) [inline arangod/StorageEngine/PhysicalCollection.h]: this will smile away all exceptions... mmmh.

File: etc/testing/arangod-common.conf [hunk 26]
```diff
@@ -1,6 +1,6 @@
 [log]
 line-number = false
-force-direct = false
+force-direct = true
```
Comment (jsteemann) [inline etc/testing/arangod-common.conf]: should not be part of this PR
```suggestion
force-direct = false
```
